cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/CQueue.scala

```diff
@@ -0,0 +1,137 @@
+package com.twitter.cassovary.util.collections
+
+import it.unimi.dsi.fastutil.PriorityQueue
+import it.unimi.dsi.fastutil.ints.{IntArrayFIFOQueue, IntComparator, IntHeapPriorityQueue,
+  IntPriorityQueue}
+import it.unimi.dsi.fastutil.objects.{ObjectArrayFIFOQueue, ObjectArrayPriorityQueue}
+
+/**
+ * Specialized implementation of queue.
+ *
+ * Implemented methods are not boxed, when using with primitive types.
+ */
+abstract class CQueue[@specialized(Int, Long) T] {
+  def first(): T
+  def +=(elem: T)
+  def deque(): T
+  def isEmpty: Boolean
+  def clear(): Unit
+  def size(): Int
+}
+
+/**
+ * FIFO [[CQueue]].
+ */
+trait FIFO[@specialized(Int, Long) T] extends CQueue[T] {
+  def enqueueFirst(elem: T)
+}
+
+object CQueue {
+  /**
+   * Creates specialized [[CQueue]] that is a FIFO queue.
+   */
+  def fifo[@specialized(Int, Long) T]()(implicit impl: CQueueFactory[T]): FIFO[T]= {
+    impl.fifo()
+  }
+
+  /**
+   * Priority [[CQueue]] that uses default ordering (implicit) for `T`.
+   */
+  def naturalPriority[@specialized(Int, Long) T]()(
```

Comment:
will the compiler allow overloaded apply() to both naturalPriority() and priority() ?